feat: add iso_oscar_singular_[coeff/poly]_ring

thofma opened the pull request:
Implements the plan we agreed on https://github.com/oscar-system/Oscar.jl/issues/976, introducing
```
    iso_oscar_singular_coeff_ring
    iso_oscar_singular_poly_ring
```

This is similar to `iso_oscar_gap`, although here we must have two separate version.

This PR here applies it only as far as we need it to fix https://github.com/oscar-system/Oscar.jl/issues/976.

Comments/thoughts:
- There are still many `singular_poly_ring` and `singular_generators` uses left, which can (and will with the right input) trigger ambiguity errors. But these might be fixed later.
- @fieker this also fixes `FqField` if the field is not absolute.
- This opens up the possibility to optimize the coercion for specific fields, so that Singular has a nice field to work with. For example:
  -  if $K$ is an arbitrary number field, we could bring it into the form $\mathbf{Q}(\alpha)$,
  - if $R = K[e]/(e^2)$, we could turn $R[x_1,\dotsc,x_n]$ into $K[e,x_1,\dotsc,x_n]/(e^2)$

CC: @fieker @ederc

thofma committed:
feat: add iso_oscar_singular_*

- use it in some places

lgoettgens commented:
The book tests are unhappy. See https://github.com/oscar-system/Oscar.jl/actions/runs/12018485868/job/33503068821?pr=4342#step:10:377

```
+ ERROR: Unexpected case in prepareQuotientring. Please inform the authors
+ leaving primdec.lib::prepareQuotientring (0)
+ leaving primdec.lib::decomp_i (3361)
+ leaving primdec.lib::primdecGTZ_i (5678)
+ 
+ Stacktrace:
+   [1] call_singular_library_procedure
+     @ ~/.julia/packages/CxxWrap/eWADG/src/CxxWrap.jl:668 [inlined]
+   [2] low_level_caller_rng(lib::String, name::String, ring::Singular.PolyRing{…}, args::Singular.sideal{…})
+     @ Singular ~/.julia/packages/Singular/JyB5B/src/caller.jl:[379](https://github.com/oscar-system/Oscar.jl/actions/runs/12018485868/job/33503068821?pr=4342#step:10:383)
+   [3] primdecGTZ(ring::Singular.PolyRing{Singular.n_FieldElem{…}}, args::Singular.sideal{Singular.spoly{…}})
+     @ Singular.LibPrimdec ~/.julia/packages/Singular/JyB5B/src/Meta.jl:45
+   [4] _compute_primary_decomposition(I::MPolyIdeal{AbstractAlgebra.Generic.MPoly{…}}; algorithm::Symbol)
+     @ Oscar ~/work/Oscar.jl/Oscar.jl/src/Rings/mpoly-ideals.jl:860
+   [5] _compute_primary_decomposition
+     @ ~/work/Oscar.jl/Oscar.jl/src/Rings/mpoly-ideals.jl:848 [inlined]
+   [6] #316
+     @ ~/work/Oscar.jl/Oscar.jl/src/Rings/mpoly-ideals.jl:818 [inlined]
+   [7] get!(default::Oscar.var\"#316#317\"{Symbol, MPolyIdeal{AbstractAlgebra.Generic.MPoly{…}}}, h::Dict{Symbol, Any}, key::Symbol)
+     @ Base ./dict.jl:479
+   [8] get_attribute!
+     @  AbstractAlgebra [inlined]
+   [9] #primary_decomposition#315
+     @ ~/work/Oscar.jl/Oscar.jl/src/Rings/mpoly-ideals.jl:817 [inlined]
+  [10] primary_decomposition
+     @ ~/work/Oscar.jl/Oscar.jl/src/Rings/mpoly-ideals.jl:815 [inlined]
+  [11] tropical_variety(I::MPolyIdeal{…}, nu::TropicalSemiringMap{…}; weighted_polyhedral_complex_only::Bool, skip_saturation::Bool, skip_primary_decomposition::Bool)
+     @ Oscar ~/work/Oscar.jl/Oscar.jl/src/TropicalGeometry/variety.jl:189
+  [12] tropical_variety(I::MPolyIdeal{…}, nu::TropicalSemiringMap{…})
+     @ Oscar ~/work/Oscar.jl/Oscar.jl/src/TropicalGeometry/variety.jl:163
+  [13] top-level scope
+     @ REPL[51]:1
+ Some type information was truncated. Use `show(err)` to see complete types.
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+"""
+    iso_oscar_singular_coeff_ring(R::Ring) -> Map
+
+Return a ring isomorphism of `R` onto a native Singular ring
```
```suggestion
Return a ring isomorphism of `R` onto a native Singular ring.
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+##############################################################################
+#
+# Conversion to and from Singular: in particular, some Rings are
```
```suggestion
# Conversion to and from Singular: in particular, some rings are
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapfpField(R, Singular.Fp(Int(characteristic(R))))
+
+# ZZ/nZZ, n small and big
+struct OscarSingularCoefficientRingMapZZMod{U} <: OscarSingularCoefficientRingMap{U, Singular.N_ZpField}
```
Isn't the second type argument the type of the codomain? And `codomain` above is set to return `f.S`. So should this be the following?
```suggestion
struct OscarSingularCoefficientRingMapZZMod{U} <: OscarSingularCoefficientRingMap{U, Singular.N_ZnRing}
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+)
+
+# QQ(a)
+struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_ZpField}
```
Same question as above: should this perhaps be
```suggestion
struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_AlgExtField}
```
If not then perhaps a comment explaining it would be helpful

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+"""
+iso_oscar_singular_poly_ring
+
+abstract type OscarSingularCoefficientRingMap{S, T} <: Map{S, T, Any, Any} end
```
Very small nitpick: I was briefly confused because `S` here is not the same as `f.S` later (i.e. `S` as two meanings). How about following the naming used in AA here, which hint sat the use, "Domain type" and "Codomain type"?
```suggestion
abstract type OscarSingularCoefficientRingMap{D, C} <: Map{D, C, Any, Any} end
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+  return OscarSingularCoefficientRingMapAbsSimpleNumField(R, SK)
+end
+
+# It is nonsense, but the conversion resides in src/number/n_algExt.jl
```
What is nonsense? Is that path relative to `Singular.jl`? Perhaps write
```suggestion
# It is nonsense, but the conversion resides in Singular.jl/src/number/n_algExt.jl
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+
+struct OscarSingularCoefficientRingMapfqPolyRepField <: OscarSingularCoefficientRingMap{fqPolyRepField, Singular.N_ZpField}
+  R::fqPolyRepField
+  S::Singular.N_AlgExtField
```
OK now I am really lost, and the comment confuses me further.... So this is a coefficient map where the codomain is declared in the type signature as having type `N_ZpField` but `S` (the actual codomain?) has type `N_AlgExtField`. But there is  comment which says this "is nonsense" so perhaps this is intending to say it is on purpose? Or what exactly is nonsense?

thofma replied:
The nonsense is that `R(S(1))` is working. The implementation is not very safe and silently gives wrong results for wrong input. The thing with the type mismatch is just a copy & paste error.

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+struct OscarSingularCoefficientRingMapFqField <: OscarSingularCoefficientRingMap{FqField, Singular.N_ZpField}
+  R::FqField
+  S#= unstable =#
+  iso::MapFromFunc{FqField, FqField}
```
The comment indicates that the type of `S` is unstable, but above the type of the codomain is declared to be `N_ZpField`. Why not do
```suggestion
struct OscarSingularCoefficientRingMapFqField{U} <: OscarSingularCoefficientRingMap{FqField, U}
  R::FqField
  S::U
  iso::MapFromFunc{FqField, FqField}
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+  S::Singular.N_ZpField
+end
+
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapfpField(R, Singular.Fp(Int(characteristic(R))))
```
Could avoid some code duplication and chance of getting types wrong by something like this (and then only introducing further subtypes in the cases needing it):
```suggestion
struct OscarSingularCoefficientRingMapGeneric{D,C} <: OscarSingularCoefficientRingMap{D,C} 
  R::D
  S::C
  OscarSingularCoefficientRingMapGeneric(R::D,S::C) where {D,C} = new{D,C}(R, S)
end

iso_oscar_singular_coeff_ring(R::ZZRing) = OscarSingularCoefficientRingMapGeneric(R, Singular.Integers())

iso_oscar_singular_coeff_ring(R::QQField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Rationals())

# prime field, small characteristic
iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Fp(Int(characteristic(R))))
```

fingolfin reviewed (commented):
Thanks a *ton* for working on this, much appreciated. I have a few questions about a few bits that genuinely confuse me. Apologies if these remarks are premature or miss the point.

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+end
+
+# catchall 
+struct OscarSingularCoefficientRingMapGeneric{U, V} <: OscarSingularCoefficientRingMap{U, V}
```
Oh so you already have this generic fallback. So I just wonder why you don't use it in more cases?

thofma replied:
I think I noticed it too late that I will need a catchall. So at the moment `*Generic` is only constructed if there is no native Singular backend. This was my interpretation of "generic". I can rewrite it as you suggest to deduplicate the code.

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+# for some reason this is used
+function image(f::OscarSingularPolyRingQuoMap, b::MPolyRingElem)
+  @assert parent(b) === base_ring(domain(f))
+  return codomain(f)(b)
```
Should this perhaps be
```suggestion
  return image(f, domain(f)(b))
```
Otherwise `f` is not used at all (other than using it to find domain and codomain)

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,458 @@
+
+function preimage(f::OscarSingularPolyRingQuoMap, a::Singular.spoly)
+  @assert parent(a) === codomain(f)
+  return MPolyQuoRingElem(preimage(f.f, a; check = false), domain(f))
```
Does this also work? Then it would seem preferable to me (avoiding the explicit type `MPolyQuoRingElem`)
```suggestion
  return domain(f)(preimage(f.f, a; check = false))
```

thofma commented:
All the confusing `N_ZpField` are copy and paste errors. Let me fix those.

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,446 @@
+  R::U
+  S::V
+end
+
```
Now `OscarSingularCoefficientRingMapGeneric` is defined twice, so gotta get rid of this one
```suggestion
```

afkafkafk13 commented:
@thofma : slightly off-topic, but directly related to your changes

I have a question relating this to issue #4303 -- and I confess that I am not too deep into the internals of the translation: 
Singular directly hardcodes a module ordering in its rings (at the moment of definition).  This is not clean at all, but it is a fact. Don't we need to take care of the module ordering translation in a suitable way to avoid inconsistencies like in #4304, (which is due to the Singular default ordering for modules being used at some points)?

thofma commented:
@fingolfin I think I addressed all your questions and this should be ready to go. I also refactored the code, so that `singular_poly_ring`/`iso_oscar_poly_ring` will stay in sync (in view of the changes in https://github.com/oscar-system/Oscar.jl/pull/4312).

@afkafkafk13 The changes here only concern how the coefficients are translated between Oscar and Singular. I don't know anything about how the modules or their interaction with Singular is designed.

afkafkafk13 commented:
> @afkafkafk13 The changes here only concern how the coefficients are translated between Oscar and Singular. I don't know anything about how the modules or their interaction with Singular is designed.

Then let us get this PR merged quickly, so that this part is already in its new structure before Christian and I meet next week. I have the suspicion that this is exactly the area where we also have to take care of existing module ordering....

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+struct OscarSingularCoefficientRingMapGeneric{D,C} <: OscarSingularCoefficientRingMap{D,C} 
+  R::D
+  S::C
+  #OscarSingularCoefficientRingMapGeneric(R::D,S::C) where {D,C} = new{D,C}(R, S)
```
```suggestion
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+  S::Singular.N_ZnRing
+end
+
+iso_oscar_singular_coeff_ring(R::Union{zzModRing, ZZModRing}) = OscarSingularCoefficientRingMapZZMod(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
```
Why can't you use `OscarSingularCoefficientRingMapGeneric` here?
```suggestion
iso_oscar_singular_coeff_ring(R::Union{zzModRing, ZZModRing}) = OscarSingularCoefficientRingMapGeneric(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+    Sminpoly += SQa(coeff(minpoly, i))*Sa^i
+  end
+  SK, _ = Singular.AlgebraicExtensionField(SQa, Sminpoly)
+  return OscarSingularCoefficientRingMapAbsSimpleNumField(R, SK)
```
Here I also wonder why you can't just use `OscarSingularCoefficientRingMapGeneric` ?
```suggestion
  return OscarSingularCoefficientRingMapGeneric(R, SK)
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+
+struct OscarSingularCoefficientRingMapfqPolyRepField <: OscarSingularCoefficientRingMap{fqPolyRepField, Singular.N_ZpField}
+  R::fqPolyRepField
+  S::Singular.N_AlgExtField
```
There is still a mismatch here between the codomain type parameter `Singular.N_ZpField` and `S::Singular.N_AlgExtField`

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+end
+
+# It is nonsense, but the conversion resides in Singular.jl/src/number/n_algExt.jl
+# via parent object call overloading
```
Still confuses me. How about this:
```suggestion
# Conversion from GF(p, n), small p, to Singular.N_AlgExtField is nonsense,
# but code for the conversion is present in Singular.jl/src/number/n_algExt.jl
# via parent object call overloading, so we wrap it here anyway.
```
Assuming this is correct, of course. If not, all the more reason to revise this ;-)

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+    Sminpoly += SFa(coeff(minpoly, i))*Sa^i
+  end
+  SF, _ = Singular.AlgebraicExtensionField(SFa, Sminpoly)
+  return OscarSingularCoefficientRingMapfqPolyRepField(F, SF)
```
```suggestion
  return OscarSingularCoefficientRingMapGeneric(F, SF)
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+  R::AbsSimpleNumField
+  S::Singular.N_AlgExtField
+end
+
```
```suggestion
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+  R::fqPolyRepField
+  S::Singular.N_AlgExtField
+end
+
```
```suggestion
```

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,448 @@
+  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), b -> begin a = Fabs(); Nemo.set!(a, b); a end)
+end
+
+# Nonsense for FqField (aka fq_default from flint)
```
What is nonsense about it? If it is nonsense why is it here? Maybe it does have a sense after all? Or maybe the comment is the nonsense? Argh, help, I can't make sense of this :-(

thofma commented:
I removed all the *sense.

thofma commented:
@fingolfin any other concerns?

fingolfin commented on src/Rings/oscar_singular.jl:
```diff
@@ -0,0 +1,432 @@
+  Kx = parent(defining_polynomial(F))
+  Fabs, QQtoFabs = Nemo._residue_field(defining_polynomial(F); absolute = true, check = false)
+  Fabsx, = polynomial_ring(Fabs, :x; cached = false)
+  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), b -> begin a = Fabs(); Nemo.set!(a, b); a end)
```
For which types does `Nemo.set!(a, b)` work but not `Fabs(b)`?

otherwise we could do
```suggestion
  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), b -> Fabs(b))
```
Or even
```suggestion
  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), Fabs)
```

thofma replied:
It is not a matter of types. This is reinterpreting elements between different flint finite fields and outside the scope of "promotion".

fingolfin reviewed (approved):
Looks good to me, merge at will
